lightning/src/onion_message/messenger.rs

```diff
@@ -304,6 +314,101 @@ where
 	}))
 }
 
+/// Decode one layer of an incoming onion message
+/// Returns either a Forward (another onion message), or Receive (decrypted content)
+pub fn peel_onion<NS: Deref, L: Deref, CMH: Deref, T: CustomOnionMessageContents>(
+	node_signer: &NS, secp_ctx: &Secp256k1<secp256k1::All>, logger: &L, custom_handler: &CMH,
```

Comment:
```suggestion
	node_signer: NS, secp_ctx: &Secp256k1<secp256k1::All>, logger: L, custom_handler: CMH,
```
Shouldn't be necessary to take references to things that are already `Deref`, IIUC.

Reply:
ok got it

Reply:
Sorry, im not sure how to call `peel_onion` from `handle_onion_message` without taking refs to `self.node_signer`, `self.logger`, and `self.custom_handler`

Reply:
Ah it was weirder than I expected actually, but this works:

```diff
diff --git a/lightning/src/onion_message/messenger.rs b/lightning/src/onion_message/messenger.rs
index 1d1512d16..1a715b7ec 100644
--- a/lightning/src/onion_message/messenger.rs
+++ b/lightning/src/onion_message/messenger.rs
@@ -317,7 +317,7 @@ where
 /// Decode one layer of an incoming onion message
 /// Returns either a Forward (another onion message), or Receive (decrypted content)
 pub fn peel_onion<NS: Deref, L: Deref, CMH: Deref, T: CustomOnionMessageContents>(
-       node_signer: &NS, secp_ctx: &Secp256k1<secp256k1::All>, logger: &L, custom_handler: &CMH,
+       node_signer: NS, secp_ctx: &Secp256k1<secp256k1::All>, logger: L, custom_handler: CMH,
        msg: &msgs::OnionMessage,
 ) -> Result<PeeledOnion<CMH>, ()>
 where
@@ -562,8 +562,13 @@ where
        /// soon we'll delegate the onion message to a handler that can generate invoices or send
        /// payments.
        fn handle_onion_message(&self, _peer_node_id: &PublicKey, msg: &msgs::OnionMessage) {
-               match peel_onion::<NS, L, CMH, <<CMH as Deref>::Target as CustomOnionMessageHandler>::CustomMessage>(
-                       &self.node_signer, &self.secp_ctx, &self.logger, &self.custom_handler, msg
+               match peel_onion::<
+                       &<NS as std::ops::Deref>::Target,
+                       &<L as std::ops::Deref>::Target,
+                       &<CMH as Deref>::Target,
+                       <<CMH as Deref>::Target as CustomOnionMessageHandler>::CustomMessage
+               >(
+                       &*self.node_signer, &self.secp_ctx, &*self.logger, &*self.custom_handler, msg
                ) {
                        Ok(PeeledOnion::Receive(message, path_id, reply_path)) => {
                                log_trace!(self.logger,

Reply:
Actually, realized you can just remove the type parameters on the `peel_onion` call:
```diff
diff --git a/lightning/src/onion_message/messenger.rs b/lightning/src/onion_message/messenger.rs
index 1d1512d16..7142e3a5c 100644
--- a/lightning/src/onion_message/messenger.rs
+++ b/lightning/src/onion_message/messenger.rs
@@ -316,8 +316,8 @@ where

 /// Decode one layer of an incoming onion message
 /// Returns either a Forward (another onion message), or Receive (decrypted content)
-pub fn peel_onion<NS: Deref, L: Deref, CMH: Deref, T: CustomOnionMessageContents>(
-       node_signer: &NS, secp_ctx: &Secp256k1<secp256k1::All>, logger: &L, custom_handler: &CMH,
+pub fn peel_onion<NS: Deref, L: Deref, CMH: Deref>(
+       node_signer: NS, secp_ctx: &Secp256k1<secp256k1::All>, logger: L, custom_handler: CMH,
        msg: &msgs::OnionMessage,
 ) -> Result<PeeledOnion<CMH>, ()>
 where
@@ -562,8 +562,8 @@ where
        /// soon we'll delegate the onion message to a handler that can generate invoices or send
        /// payments.
        fn handle_onion_message(&self, _peer_node_id: &PublicKey, msg: &msgs::OnionMessage) {
-               match peel_onion::<NS, L, CMH, <<CMH as Deref>::Target as CustomOnionMessageHandler>::CustomMessage>(
-                       &self.node_signer, &self.secp_ctx, &self.logger, &self.custom_handler, msg
+               match peel_onion(
+                       &*self.node_signer, &self.secp_ctx, &*self.logger, &*self.custom_handler, msg
                ) {
                        Ok(PeeledOnion::Receive(message, path_id, reply_path)) => {
                                log_trace!(self.logger,
```

Reply:
oh, that is much cleaner, thank u!